### PR TITLE
config: Change xds-server-version to be plural and serve multiple xDS versions

### PR DESCRIPTION
**Pull request by stevesloka**
Changes the xds-server-version config item to be plural so that users can specify a set
of xDS versions Contour will serve. Defaults to V2 only for now.

Updates #1898
Updates #3047 

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: pkg/config/parameters.go [hunk 1]
```diff
@@ -333,8 +333,13 @@ func (p *Parameters) Validate() error {
 		return err
 	}
 
-	if err := p.Server.XDSServerVersion.Validate(); err != nil {
-		return err
+	if len(p.Server.XDSServerVersions) == 0 {
+		return fmt.Errorf("at least one xds server version must be specified")
+	}
+	for _, v := range p.Server.XDSServerVersions {
```
Comment (skriss) [inline pkg/config/parameters.go:339]: it'd probably be good to de-dupe the list too, so `[v2, v2]` becomes just `[v2]`.

**Review by skriss — APPROVED** [review]
This looks OK to me pending the one comment.

This field will default to `[v2]` now, then `[v2, v3]` when we support both, and then just `[v3]` once we drop `v2` support - right?

File: internal/xds/server.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+)
+
+// If registry is non-nil gRPC server metrics will be automatically
+// configured and enabled.
+func NewServer(registry *prometheus.Registry, opts ...grpc.ServerOption) *grpc.Server {
```
Comment (jpeach) [inline internal/xds/server.go:24]: nit: we could make the `registry` argument compulsory and just use `grpc.NewServer` when there are no metrics. Not a big deal either way though.

**Review by jpeach — APPROVED** [review]
LGTM, just a few nits. Nice work!

**Comment by stevesloka** [comment]
> This field will default to [v2] now, then [v2, v3] when we support both, and then just [v3] once we drop v2 support - right?

Yes, once we fully make v3 the default, we'll swap the default values, then will need to deprecate out the v2 version.

File: cmd/contour/serve.go [hunk 3]
```diff
@@ -652,32 +652,30 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
 		case config.EnvoyServerType:
-			switch ctx.Config.Server.XDSServerVersion {
-			case config.XDSv3:
-				log.Fatalf("xDS server type %q not yet implemented for server type %q!", ctx.Config.Server.XDSServerVersion, ctx.Config.Server.XDSServerType)
-			default:
-				grpcServer = contour_xds_v2.RegisterServer(
-					envoy_server_v2.NewServer(context.Background(), snapshotCache, nil),
-					registry,
-					ctx.grpcOptions(log)...)
+			grpcServer = xds.NewServer(registry, ctx.grpcOptions(log)...)
```
[the remark this answers is not in the document]
Comment (stevesloka) [reply cmd/contour/serve.go:656]: ahh good call!

File: cmd/contour/serve.go [hunk 4]
```diff
@@ -652,32 +652,30 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
+				case config.XDSv2:
+					contour_xds_v2.RegisterServer(envoy_server_v2.NewServer(context.Background(), snapshotCache, nil), grpcServer)
+				case config.XDSv3:
+					log.Fatalf("xDS server type %q not yet implemented for server type %q!", ver, ctx.Config.Server.XDSServerType)
+				}
 			}
 		default:
-			// XXX(jpeach) can't happen due to config validation.
-			log.Fatalf("invalid xdsServerType %q configured", ctx.Config.Server.XDSServerType)
+			grpcServer = xds.NewServer(registry, ctx.grpcOptions(log)...)
```
[the remark this answers is not in the document]
Comment (stevesloka) [reply cmd/contour/serve.go:667]: ok I can add back. We should already get the error message earlier that that value is bad and panic, but agree it makes sense to be explicit.

[the remark this answers is not in the document]
Comment (stevesloka) [reply cmd/contour/serve.go:667]: @jpeach I'm going to add the same `default` condition for the version as well to keep that consistent.

File: internal/xds/server.go (hunk 2 above)
Comment (stevesloka) [reply internal/xds/server.go:24]: Yeah I think that makes sense. I have an issue covering that I think that we could do that work there.

**Comment by stevesloka** [comment]
@jpeach all updated. Also please let me know if you agree with the extra default cases for the version:

-https://github.com/stevesloka/contour/blob/bb60880a172875024a276914e7a5a83c8e331285/cmd/contour/serve.go#L662
- https://github.com/stevesloka/contour/blob/bb60880a172875024a276914e7a5a83c8e331285/cmd/contour/serve.go#L675

File: cmd/contour/serve.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply cmd/contour/serve.go]: Sounds good!

File: cmd/contour/serve.go [hunk 5]
```diff
@@ -649,31 +648,32 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
 		case config.EnvoyServerType:
-			switch ctx.Config.Server.XDSServerVersion {
-			case config.XDSv3:
-				log.Fatalf("xDS server type %q not yet implemented for server type %q!", ctx.Config.Server.XDSServerVersion, ctx.Config.Server.XDSServerType)
-			default:
-				grpcServer = contour_xds_v2.RegisterServer(
-					envoy_server_v2.NewServer(context.Background(), snapshotCache, nil),
-					registry,
-					ctx.grpcOptions(log)...)
+			for _, ver := range ctx.Config.Server.XDSServerVersions {
```
Comment (jpeach) [inline cmd/contour/serve.go:655]: Quick drive-by suggestion ... it might be cleaner to make this more like `xds.NewContourServer(vers, log, other params)` and let the package deal with the versioning internally.

File: internal/featuretests/v2/featuretests.go [hunk 6]
```diff
@@ -130,9 +131,8 @@ func setup(t *testing.T, opts ...interface{}) (cache.ResourceEventHandler, *Cont
 	l, err := net.Listen("tcp", "127.0.0.1:0")
 	require.NoError(t, err)
 
-	srv := contour_xds_v2.RegisterServer(
-		contour_xds_v2.NewContourServer(log, xdscache.ResourcesOf(resources)...),
-		r /* Prometheus registry */)
+	srv := xds.NewServer(r /* Prometheus registry */)
```
Comment (skriss) [inline internal/featuretests/v2/featuretests.go:134]: we should probably rename `r` and drop the comment :)

File: cmd/contour/serve.go (hunk 5 above)
Comment (stevesloka) [reply cmd/contour/serve.go:655]: I had that originally, but ran into import cycle errors. 😢

File: pkg/config/parameters.go (hunk 1 above)
Comment (stevesloka) [reply pkg/config/parameters.go]: I added a bit about de-duping in the validate method. Please take a look @skriss

**Review by skriss — APPROVED** [review]
LGTM
